forc-pkg/src/pkg.rs

```diff
@@ -1513,19 +1500,26 @@ pub const CONTRACT_ID_CONSTANT_NAME: &str = "CONTRACT_ID";
 ///
 /// This function also ensures that if `std` exists in the graph,
 /// then the std prelude will also be added.
+///
+/// `contract_id_value` should only be Some when producing the `dependency_namespace` for a contract with tests enabled.
+/// This allows us to provide a contract's `CONTRACT_ID` constant to its own unit tests.
 pub fn dependency_namespace(
     lib_namespace_map: &HashMap<NodeIx, namespace::Module>,
     compiled_contract_deps: &CompiledContractDeps,
     graph: &Graph,
     node: NodeIx,
-    constants: BTreeMap<String, ConfigTimeConstant>,
     engines: Engines<'_>,
+    contract_id_value: Option<ContractIdConst>,
```

Comment:
If you prefer removing the `Option` in the function params we can introduce something like:

```rust
enum NamespaceKind {
  Build,
  Test(ContractIdConst)
}
```
and then we can refactor `pub fn dependency_namespace()` to 

```rust
pub fn dependency_namespace(
  lib_namespace_map: &HashMap<NodeIx, namespace::Module>,
  compiled_contract_deps: &CompiledContractDeps,
  graph: &Graph,
  node: NodeIx,
  engines: Engines<'_>,
  namespace_type: NamespaceKind
)
```

and match on `NamepsaceKind::Test(contract_id)` like:

```rust
let mut namespace = if let NamepsaceKind::Test(contract_id_value) = namespace_type {
  namespace::Module::default_with_contract_id(engines, name.clone(), contract_id_value)?
} else {
  namespace::Module::default()
};
```

Reply:
I'm thinking we can maybe do without this since inputting the contract ID _might_ be a temporary thing we do here :thinking: if it's the case that my followup PR to tackle #3077 does not change the way we feed the contract ID into the namespace, i'll implement this approach